url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/domain/singleagent/sokoban2/Sokoban2Parser.java

```diff
@@ -0,0 +1,120 @@
+package domain.singleagent.sokoban2;
+
+import java.util.List;
+
+import burlap.oomdp.auxiliary.StateParser;
+import burlap.oomdp.core.Domain;
+import burlap.oomdp.core.ObjectInstance;
+import burlap.oomdp.core.State;
+
+public class Sokoban2Parser implements StateParser {
+
+	protected Domain		domain;
+	
+	public Sokoban2Parser(Domain domain) {
+		this.domain = domain;
+	}
+	
+	@Override
+	public String stateToString(State s) {
+		
+		StringBuffer buf = new StringBuffer();
+		List<ObjectInstance> rooms = s.getObjectsOfTrueClass(Sokoban2Domain.CLASSROOM);
+		List<ObjectInstance> doors = s.getObjectsOfTrueClass(Sokoban2Domain.CLASSDOOR);
+		List<ObjectInstance> blocks = s.getObjectsOfTrueClass(Sokoban2Domain.CLASSBLOCK);
+		List<ObjectInstance> agents = s.getObjectsOfTrueClass(Sokoban2Domain.CLASSAGENT);
+
+		for(ObjectInstance o : rooms){
+			buf.append("room,");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTCOLOR)).append(",");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTLEFT)).append(",");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTTOP)).append(",");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTRIGHT)).append(",");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTBOTTOM)).append(" ");
+		}
+		for(ObjectInstance o : doors){
+			buf.append("door,");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTLEFT)).append(",");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTTOP)).append(",");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTRIGHT)).append(",");
+			buf.append(o.getDiscValForAttribute(Sokoban2Domain.ATTBOTTOM)).append(" ");
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 35:41, 42:48.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.